### PR TITLE
Improvement of Scatter Kernels

### PR DESCRIPTION
**Pull request by fabianlim**
This PR:
- adds guards for `torch<2.4` where`torch.library.customop`.
- adds more `triton.Config` for `groupXtoY` which sees better throughputs 
- `triton.autotune` needs to have the keys (e.g., `N, K`) to be `constaexpr` so that it can tune properly. Adjusted that in various functions.

### Performance

The kernel that was of the most concern was the `groupXtoY` kernel, as it required an inner iteration that would depend on the seq length `M`. We had done some tests to see what would be the recommend tuning
- All tests below do not involve compile.

#### Choice of Tunable Configs
- we prevent one option of 3 main configs, and a second one which adds 3 more on top (for a total of 6)
- before this PR, there was only one config
- we found that 3 base configs were a good tradeoff in terms of stablity and performance. 
- the 6 configs case had extreme degradation for the picture on the lower left.

image

### Tests

Tested on A100, without compile. All tests pass except 1 with a very small margin on error.

```
self = <tests.scattermoe_test.ScatterMoETest testMethod=test_scattermoe_triton_05>, x = tensor([ 0.3474, -1.1307, -1.7000,  ...,  0.3483, -0.8569, -0.5135],
       device='cuda:0')
y = tensor([ 0.3472, -1.1326, -1.7025,  ...,  0.3488, -0.8588, -0.5148],
       device='cuda:0'), exact_match = False, rtol_float32 = 0, atol_float32 = 0.006, rtol_float16 = 0, atol_float16 = 0.004, rtol_bfloat16 = 0
atol_bfloat16 = 0.04

    def assert_equal_tensors(
        self,
        x: torch.Tensor,
        y: torch.Tensor,
        exact_match: bool,
        rtol_float32: float = None,
        atol_float32: float = None,
        rtol_float16: float = None,
        atol_float16: float = None,
        rtol_bfloat16: float = None,
        atol_bfloat16: float = None,
    ) -> None:
        if exact_match:
            assert x.equal(y)
        else:
            assert x.dtype == y.dtype
            dtype = x.dtype

            if dtype == torch.float32:
>               assert_close(x, y, rtol=rtol_float32, atol=atol_float32)
E               AssertionError: Tensor-likes are not close!
E
E               Mismatched elements: 1 / 2048 (0.0%)
E               Greatest absolute difference: 0.0062046051025390625 at index (298,) (up to 0.006 allowed)
E               Greatest relative difference: 0.002081993967294693 at index (298,) (up to 0 allowed)

tests/test_commons.py:68: AssertionError
```

### CODE REVIEW

File: khd/kernels/scattermoe/triton_implementation/kernels.py [hunk 1]
```diff
@@ -113,8 +123,9 @@ def groupXtY_triton_kernel(
     stride_dwk,
     stride_dwn,
     expert_offsets_ptr,
-    K,
-    N,
+    K: tl.constexpr,
+    N: tl.constexpr,
+    E: tl.constexpr,
```
Comment (mayank31398) [inline khd/kernels/scattermoe/triton_implementation/kernels.py:128]: looks like E is not used in this kernel, lets not pass this.

File: khd/kernels/scattermoe/triton_implementation/kernels.py [hunk 2]
```diff
@@ -26,9 +26,9 @@ def scatter2scatter_triton_kernel(
     block_start_idx_ptr,
     FAN_OUT,
     M,
```
Comment (mayank31398) [inline khd/kernels/scattermoe/triton_implementation/kernels.py:28]: M should be also constexpr here?
to make autotune work with this?

**Review by mayank31398 — COMMENTED** [review]
small changes

File: khd/kernels/scattermoe/triton_implementation/kernels.py (hunk 2 above)
Comment (fabianlim) [reply khd/kernels/scattermoe/triton_implementation/kernels.py:28]: icic good catch. actually `M` changes alot so im not sure if we should remove it from `autotune`

File: khd/kernels/scattermoe/triton_implementation/kernels.py (hunk 1 above)
Comment (fabianlim) [reply khd/kernels/scattermoe/triton_implementation/kernels.py:128]: `E` is used in autotune, i added it see above

File: khd/kernels/scattermoe/triton_implementation/kernels.py (hunk 2 above)
Comment (mayank31398) [reply khd/kernels/scattermoe/triton_implementation/kernels.py:28]: yes agree

**Pull request merged**